src/main/kotlin/com/openlattice/data/storage/IndexingMetadataManager.kt

```diff
@@ -1,34 +1,83 @@
+
+                entities.entries
+                        .groupBy({ getPartition(it.key, partitions) }, { it.toPair() })
+                        .mapValues { it.value.toMap() }
+                        .forEach { (partition, entitiesWithLastWrite) ->
+
+                            entitiesWithLastWrite.entries
+                                    .groupBy { it.value }
+                                    .mapValues { it.value.map { it.key } }
+                                    .forEach { (lastWrite, entities) ->
+                                        count += markAsIndexed(connection, entitySetId, entities, partition, lastWrite)
+                                    }
+                        }
+            }
+        }
+        return count
     }
 
-    fun markAsIndexed(entityKeyIdsWithLastWrite: Map<UUID, Map<UUID, OffsetDateTime>>, linking: Boolean): Int {
+    private fun markAsIndexed(
+            connection: Connection,
+            entitySetId: UUID,
+            entityKeyIds: List<UUID>,
+            partition: Int,
+            lastWrite: OffsetDateTime
+    ): Int {
+        return connection.prepareStatement(updateLastIndexSql).use { stmt ->
+            val entitiesArray = PostgresArrays.createUuidArray(connection, entityKeyIds)
+            stmt.setObject(1, lastWrite)
+            stmt.setObject(2, entitySetId)
+            stmt.setArray(3, entitiesArray)
+            stmt.setInt(4, partition)
+            stmt.executeUpdate()
+        }
+    }
+
+    fun markLinkingIdsAsIndexed(linkingEntityKeyIdsWithLastWrite: Map<UUID, Map<UUID, OffsetDateTime>>): Int {
         return hds.connection.use { connection ->
-            val updateSql = updateLastIndexSql(linking)
-            connection.prepareStatement(updateSql).use { stmt ->
-                entityKeyIdsWithLastWrite.forEach { (entitySetId, entities) ->
-                    entities.forEach { (entityId, lastWrite) ->
-                        stmt.setObject(1, entitySetId)
-                        stmt.setObject(2, entityId)
-                        stmt.setObject(3, lastWrite)
-                        stmt.addBatch()
-                    }
+            val entitySetPartitions = partitionManager
+                    .getEntitySetsPartitionsInfo(linkingEntityKeyIdsWithLastWrite.keys)
+
+            connection.prepareStatement(updateLastLinkingIndexSql).use { stmt ->
+                linkingEntityKeyIdsWithLastWrite.forEach { (entitySetId, linkingEntities) ->
+                    val partitions = entitySetPartitions.getValue(entitySetId).partitions
+                    val partitionsArray = PostgresArrays.createIntArray(connection, partitions)
+
+                    linkingEntities.entries
+                            .groupBy { it.value }
+                            .mapValues { it.value.map { it.key } }
+                            .forEach { (lastWrite, linkingIds) ->
```

Comment:
Is this param actually lastWrite or is it lastIndex?

Reply:
We update last_index with the value of last_write. So here I would still use lastWrite, since it's the value of that

Reply:
ohhh that makes sense

Reply:
We agreed on using last write instead of now.. just don't remembering why exactly.. "Load last write during indexing and once indexing is done, set last index to last write instead of index => now()"
Can we derive the partition from linking ids as well?

Reply:
FYI: we only use this call for a single linkingId right now

Reply:
Ahh.. totally confused old entity_key_ids table with new ids table with separate rows for linking entities. Will make the change!